cpp/oneapi/dal/algo/decision_tree/common.cpp

```diff
@@ -0,0 +1,228 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/decision_tree/backend/node_info_impl.hpp"
+#include "oneapi/dal/detail/error_messages.hpp"
+
+namespace oneapi::dal::decision_tree {
+namespace v1 {
+
+namespace de = dal::detail;
+
+template <typename Task>
+node_info<Task>::node_info() : impl_(new impl_t{}) {}
+
+template <typename Task>
+node_info<Task>::~node_info() {
+    if (impl_)
+        delete impl_;
+}
+
+template <typename Task>
+node_info<Task>::node_info(impl_t* impl) : impl_(impl) {}
+
+template <typename Task>
+node_info<Task>::node_info(const node_info<Task>& orig) : impl_(new impl_t(*orig.impl_)) {}
+
+template <typename Task>
+node_info<Task>::node_info(node_info<Task>&& orig) : impl_(orig.impl_) {
+    orig.impl_ = nullptr;
+}
+
+template <typename Task>
+node_info<Task>& node_info<Task>::operator=(const node_info<task_t>& orig) {
+    if (&orig == this)
+        return *this;
+
+    if (impl_)
+        delete impl_;
```

Comment:
This check can be omitted, the same for others

Reply:
fixed